src/formats/SMI.cpp

```diff
@@ -0,0 +1,737 @@
+/// \param previous_atom Atom which was bound to the current atom
+static void find_rings_helper(
+    const std::vector<std::vector<size_t>>& adj_list,
+    std::vector<bool>& hit_atoms, std::set<Bond>& ring_bonds,
+    std::unordered_map<size_t, size_t>& ring_atoms,
+    size_t current_atom, size_t previous_atom) {
+
+    auto& current_atom_bonds = adj_list[current_atom];
+    hit_atoms[current_atom] = true;
+
+    for (auto neighbor : current_atom_bonds) {
+        // prevent back tracking
+        if (neighbor == previous_atom) {
+            continue;
+        }
+
+        // We've seen this neighbor before! Ring found
+        if (hit_atoms[neighbor]) {
+            // Don't process the same ring connection twice
+            if (ring_bonds.count(Bond(neighbor, current_atom))) {
+                continue;
+            }
+            ring_bonds.insert(Bond(neighbor, current_atom));
+
+            auto neigh_ring_iter = ring_atoms.find(neighbor);
+            if (neigh_ring_iter == ring_atoms.end()) {
+                neigh_ring_iter = ring_atoms.insert({neighbor, 0}).first;
+            }
+            ++neigh_ring_iter->second;
+
+            continue; // No need to see this atom again
+        }
+
+        // Continue the search
+        find_rings_helper(adj_list, hit_atoms, ring_bonds,
+                          ring_atoms, neighbor, current_atom);
+    }
+}
+
+static void find_rings(
+    const std::vector<std::vector<size_t>>& adj_list,
+    std::unordered_map<size_t, size_t>& ring_atoms) {
+
+    ring_atoms.clear();
+    std::vector<bool> hit_atoms(adj_list.size(), false);
+    std::set<Bond> ring_bonds;
+    
+    while (!all(hit_atoms)) {
+        auto not_hit = std::find(hit_atoms.begin(), hit_atoms.end(), false);
+        auto current_atom = static_cast<size_t>(std::distance(hit_atoms.begin(), not_hit));
+        auto& current_atom_bonds = adj_list[current_atom];
+        hit_atoms[current_atom] = true;
+        if (current_atom_bonds.empty()) {
+            continue;
+        }
+
+        find_rings_helper(adj_list, hit_atoms, ring_bonds, ring_atoms,
+                          current_atom_bonds[0], current_atom
+        );
+    }
+}
+
+/// Sets int_part to `value` if `value` is actually an integer
+static bool double_to_int(double value, int& int_part) {
+    double integer_section;
+    auto remainder = std::modf(value, &integer_section);
+    if (remainder == 0.0) {
+        int_part = static_cast<int>(std::floor(integer_section));
+        return true;
+    }
+    return false;
+}
+
+static void write_atom_smiles(std::unique_ptr<TextFile>& file_, const Atom& atom) {
+    bool needs_brackets = false;
+    auto type = atom.type();
+
+    // The mass must be an integer is the only check we need as all atoms in the
+    // periodic table have non-integer masses. Therefore, if the the mass is
+    // an integer, then we know the user has set an isotope.
+    int mass_int = 0;
+    if (double_to_int(atom.mass(), mass_int)) {
```

Comment:
If the mass is say `12.5`, this will completely ignore it, right?

Reply:
Correct. This format does not support non-integer masses.

Reply:
Should we have a warning for this? The unfortunate issue I can see here is that we will set non-integer masses when creating Atom (see: https://github.com/chemfiles/chemfiles/blob/bc375d2d5fdc613b9085f31920946598ab7ebe01/src/periodic_table.cpp#L13 we set the hydrogen mass to 1.008). 

At the same time we don't want to output a mass of 1 for all hydrogen atoms, but we don't want specific masses set by the users to be ignored silently.

Reply:
Many formats do not support floating point masses (SDF/CML/MOl2/HIN comes to mind) so I think there should be a larger PR addressing this issue.

Reply:
Ok, maybe open an issue so that we remember to do it?